File: history/day_16.ts
import { Logger, Part, run, Type } from "../day_utils"
import { World2D } from "../map2d.utils";
import { generator } from "../utils";
enum CellType {
    MIRROR_SLASH = '/',
    MIRROR_BACKSLASH = "\\",
    SPLITTER_HORIZONTAL = "-",
    SPLITTER_VERTICAL = "|",
    EMPTY = "."
}

interface Cell {
    type: CellType,
    is_empowered: boolean,
    casted_rays: Ray[]
}

interface Ray {
    startPos: World2D.Pos,
    dir: World2D.Dir
}

class Day16Map extends World2D.Map2d<Cell>{
    private nb_empowered: number = 0;
    constructor(cells: Cell[][]) {
        super(cells)
    }

    private manage_empower(c: Cell) {
        if (!c.is_empowered) {
            c.is_empowered = true;
            this.nb_empowered++;
        }
    }

    public get_next_ray(new_pos: World2D.Pos, dir: World2D.Dir): Ray[] {
        let potential_rays: Ray[] = [];
        const c = this.cell(new_pos);
        if (c.type === CellType.SPLITTER_HORIZONTAL) {
            this.apply_in_dir(new_pos, World2D.Dir.LEFT, (c, p, d) => {
                potential_rays.push({ dir: d, startPos: p });
            });
            this.apply_in_dir(new_pos, World2D.Dir.RIGHT, (c, p, d) => {
                potential_rays.push({ dir: d, startPos: p });
            });
        } else if (c.type === CellType.SPLITTER_VERTICAL) {
            this.apply_in_dir(new_pos, World2D.Dir.UP, (c, p, d) => {
                potential_rays.push({ dir: d, startPos: p });
            });
            this.apply_in_dir(new_pos, World2D.Dir.DOWN, (c, p, d) => {
                potential_rays.push({ dir: d, startPos: p });
            });
        } else if (c.type === CellType.MIRROR_BACKSLASH) {
            switch (dir) {
                case World2D.Dir.LEFT: this.apply_in_dir(new_pos, World2D.Dir.UP, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.RIGHT: this.apply_in_dir(new_pos, World2D.Dir.DOWN, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.UP: this.apply_in_dir(new_pos, World2D.Dir.LEFT, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.DOWN: this.apply_in_dir(new_pos, World2D.Dir.RIGHT, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
            }
        } else if (c.type === CellType.MIRROR_SLASH) {
            switch (dir) {
                case World2D.Dir.RIGHT: this.apply_in_dir(new_pos, World2D.Dir.UP, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.LEFT: this.apply_in_dir(new_pos, World2D.Dir.DOWN, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.UP: this.apply_in_dir(new_pos, World2D.Dir.RIGHT, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
                case World2D.Dir.DOWN: this.apply_in_dir(new_pos, World2D.Dir.LEFT, (c, p, d) => {
                    potential_rays.push({ dir: d, startPos: p });
                }); break
            }
        }
        const to_cast = potential_rays.filter(r => c.casted_rays.find(cr => this.is_same_pos(r.startPos, cr.startPos)) === undefined);
        to_cast.forEach(tc => c.casted_rays.push(tc));

        return to_cast
    }

    public empoweredView(): string[] {
        return this.toStringArray((c) => c.is_empowered ? "#" : ".");
    }

    public shoot_ray(ray: Ray): Ray[] {
        const start_cell = this.cell(ray.startPos);
        this.manage_empower(start_cell);
        if (start_cell.type !== CellType.EMPTY) {
            return this.get_next_ray(ray.startPos, ray.dir);
        }

        const next_pos = this.move_while(ray.startPos, ray.dir, (c) => {
            this.manage_empower(c);
            return c.type === CellType.EMPTY;
        })
        if (next_pos === undefined) {
            return [];
        }

        return this.get_next_ray(next_pos, ray.dir);
    }

    public empower(start: Ray): number {
        let rays: Ray[] = [start];
        while (rays.length > 0) {
            rays = rays.flatMap(ray => this.shoot_ray(ray));
        }
        return this.nb_empowered;
    }
}

function parse(lines: string[]): Day16Map {
    const map = lines.map(l => l.split('').map(t => {
        let cell: Cell = { type: t as CellType, is_empowered: false, casted_rays: [] };
        return cell;
    }));
    return new Day16Map(map);
}


function puzzle(lines: string[], part: Part, type: Type, logger: Logger): void {
    const data = parse(lines);
    if (part === Part.PART_1) {
        const result = data.empower({ startPos: { x: 0, y: 0 }, dir: World2D.Dir.RIGHT });
        logger.debug(() => ["Result :\n"].concat(data.empoweredView()));
        logger.result(result, [46, 7951])
    }
    else {
        let max = -1;
        const clone_cell: World2D.Clone<Cell> = (c) => { return { ...c, casted_rays: [] } };
        max = Math.max(new Day16Map(data.cloned_cells(clone_cell)).empower({ startPos: { x: 0, y: 0 }, dir: World2D.Dir.RIGHT }), max);
        [...generator(data.height())].forEach((y: number) => {
            max = Math.max(new Day16Map(data.cloned_cells(clone_cell)).empower({ startPos: { x: 0, y }, dir: World2D.Dir.RIGHT }), max);
            max = Math.max(new Day16Map(data.cloned_cells(clone_cell)).empower({ startPos: { x: data.width() - 1, y }, dir: World2D.Dir.LEFT }), max);
        });

        [...generator(data.width())].forEach((x) => {
            max = Math.max(new Day16Map(data.cloned_cells(clone_cell)).empower({ startPos: { x: x, y: 0 }, dir: World2D.Dir.DOWN }), max);
            max = Math.max(new Day16Map(data.cloned_cells(clone_cell)).empower({ startPos: { x: x, y: data.height() - 1 }, dir: World2D.Dir.UP }), max);
        });
        logger.result(max, [51, 8148])
    }
}

/**
 * Update the date number after copy
 * Adapt types list to your needs and parts also 
 */
run(16, [Type.TEST, Type.RUN], puzzle, [Part.PART_1, Part.PART_2])

File: rust/src/day16.rs
use std::hash::Hasher;

use rustc_hash::FxHashSet;

use crate::{
    check_result, log,
    map2d::{Direction, Map2D, Pos},
    utils::Context,
};

#[derive(Debug, PartialEq, Eq)]
enum CellType {
    MirrorSlash,
    MirrorBackslash,
    SplitterHorizontal,
    SplitterVertical,
    Empty,
}

impl CellType {
    fn from_char(c: char) -> CellType {
        match c {
            '/' => CellType::MirrorSlash,
            '\\' => CellType::MirrorBackslash,
            '-' => CellType::SplitterHorizontal,
            '|' => CellType::SplitterVertical,
            '.' => CellType::Empty,
            _ => panic!("Cannot map {}", c),
        }
    }

    fn next_dirs(&self, orig_dir: &Direction) -> Vec<Direction> {
        match self {
            CellType::SplitterVertical => match orig_dir {
                Direction::DOWN => vec![Direction::DOWN],
                Direction::UP => vec![Direction::UP],
                Direction::LEFT | Direction::RIGHT => vec![Direction::DOWN, Direction::UP],
            },
            CellType::SplitterHorizontal => match orig_dir {
                Direction::DOWN | Direction::UP => vec![Direction::LEFT, Direction::RIGHT],
                Direction::LEFT => vec![Direction::LEFT],
                Direction::RIGHT => vec![Direction::RIGHT],
            },
            CellType::MirrorBackslash => match orig_dir {
                Direction::DOWN => vec![Direction::RIGHT],
                Direction::LEFT => vec![Direction::UP],
                Direction::RIGHT => vec![Direction::DOWN],
                Direction::UP => vec![Direction::LEFT],
            },
            CellType::MirrorSlash => match orig_dir {
                Direction::DOWN => vec![Direction::LEFT],
                Direction::LEFT => vec![Direction::DOWN],
                Direction::RIGHT => vec![Direction::UP],
                Direction::UP => vec![Direction::RIGHT],
            },
            CellType::Empty => panic!("Should not occurs"),
        }
    }
}

#[derive(Debug)]
struct Cell {
    c_type: CellType,
    potential_rays: [Option<Ray>; 4], //To Directions UP, DOWN, LEFT, RIGHT
}

impl Cell {
    fn get_potential_ray_pos(dir: &Direction) -> usize {
        match dir {
            Direction::UP => 0,
            Direction::DOWN => 1,
            Direction::LEFT => 2,
            Direction::RIGHT => 3,
        }
    }

    fn get_potential_ray<'a>(&'a self, dir: &Direction) -> Option<&'a Ray> {
        self.potential_rays[Cell::get_potential_ray_pos(dir)].as_ref()
    }

    fn build_applicable_rays(&self, map: &Map2D<Cell>, pos: &Pos) -> [Option<Ray>; 4] {
        match self.c_type {
            CellType::Empty => [
                if map.is_down_border(pos) {
                    Ray::new_ray(map, pos, &Direction::UP)
                } else {
                    None
                },
                if map.is_up_border(pos) {
                    Ray::new_ray(map, pos, &Direction::DOWN)
                } else {
                    None
                },
                if map.is_right_border(pos) {
                    Ray::new_ray(map, pos, &Direction::LEFT)
                } else {
                    None
                },
                if map.is_left_border(pos) {
                    Ray::new_ray(map, pos, &Direction::RIGHT)
                } else {
                    None
                },
            ],
            CellType::MirrorBackslash | CellType::MirrorSlash => [
                Ray::new_ray(map, pos, &Direction::UP),
                Ray::new_ray(map, pos, &Direction::DOWN),
                Ray::new_ray(map, pos, &Direction::LEFT),
                Ray::new_ray(map, pos, &Direction::RIGHT),
            ],
            CellType::SplitterVertical => [
                Ray::new_ray(map, pos, &Direction::UP),
                Ray::new_ray(map, pos, &Direction::DOWN),
                None,
                None,
            ],
            CellType::SplitterHorizontal => [
                None,
                None,
                Ray::new_ray(map, pos, &Direction::LEFT),
                Ray::new_ray(map, pos, &Direction::RIGHT),
            ],
        }
    }
}

#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
enum ToMark {
    Vertical(Pos, usize),
    Horizontal(Pos, usize),
}

impl ToMark {
    fn new(pos_a: &Pos, pos_b: &Pos) -> ToMark {
        if pos_a.x == pos_b.x {
            let delta_y = pos_a.y.abs_diff(pos_b.y) + 1;
            if pos_a.y < pos_b.y {
                ToMark::Vertical(pos_a.clone(), delta_y)
            } else {
                ToMark::Vertical(pos_b.clone(), delta_y)
            }
        } else {
            let delta_x = pos_a.x.abs_diff(pos_b.x) + 1;
            if pos_a.x < pos_b.x {
                ToMark::Horizontal(pos_a.clone(), delta_x)
            } else {
                ToMark::Horizontal(pos_b.clone(), delta_x)
            }
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
struct Ray {
    start_pos: Pos,
    dir: Direction,
    end_pos: Pos,
    is_outgoing: bool,
    to_mark: ToMark,
}

impl std::hash::Hash for Ray {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.start_pos.hash(state);
        self.dir.hash(state);
    }
}

impl Ray {
    fn new_ray(map: &Map2D<Cell>, pos: &Pos, dir: &Direction) -> Option<Ray> {
        let target_pos = map
            .iter_dir(*pos, *dir, false)
            .skip_while(|curr_pos| map.get(curr_pos).c_type == CellType::Empty)
            .nth(0)
            .unwrap_or_else(|| map.move_to_border(pos, dir));

        if target_pos == *pos {
            return None;
        } else {
            return Some(Ray {
                to_mark: ToMark::new(&pos, &target_pos),
                dir: *dir,
                start_pos: *pos,
                end_pos: target_pos,
                is_outgoing: map.get(&target_pos).c_type == CellType::Empty,
            });
        }
    }
}

type InputMap = Map2D<Cell>;

fn parse(lines: &Vec<String>) -> InputMap {
    let content = lines
        .iter()
        .map(|l| {
            l.chars()
                .map(|c| Cell {
                    c_type: CellType::from_char(c),
                    potential_rays: [None; 4],
                })
                .collect::<Vec<Cell>>()
        })
        .collect::<Vec<Vec<Cell>>>();

    let mut map = Map2D::new(content);
    for any_pos in map.iter_all_fast() {
        let rays = map.get(&any_pos).build_applicable_rays(&map, &any_pos);
        map.get_mut(&any_pos).potential_rays = rays;
    }
    return map;
}

struct Solution<'a> {
    casted_rays: FxHashSet<&'a Ray>,
    marks: FxHashSet<&'a ToMark>,
}

fn non_empty_cell_next_dirs<'a>(cell: &'a Cell, orig_dir: &Direction) -> Vec<&'a Ray> {
    cell.c_type
        .next_dirs(orig_dir)
        .iter()
        .filter_map(|dir| cell.get_potential_ray(dir))
        .collect()
}

fn manage_ray_cast<'a>(map: &'a InputMap, ray: &'a Ray, sol: &mut Solution<'a>) -> Vec<&'a Ray> {
    if sol.casted_rays.contains(&ray) {
        return vec![];
    }
    sol.marks.insert(&ray.to_mark);
    sol.casted_rays.insert(&ray);
    if ray.is_outgoing {
        return vec![];
    }
    let cell = map.get(&ray.end_pos);

    return non_empty_cell_next_dirs(cell, &ray.dir);
}

fn to_string(resulting: &Map2D<u32>) -> String {
    resulting
        .get_content()
        .iter()
        .map(|l| l.iter().map(|v| if *v > 0 { '#' } else { '.' }).collect::<String>())
        .collect::<Vec<String>>()
        .join("\n")
}

fn solve(map: &InputMap, start_pos: Pos, dir: &Direction, context: &Context) -> usize {
    let mut solution = Solution {
        casted_rays: FxHashSet::default(),
        marks: FxHashSet::default(),
    };

    let mut next_to_process: Vec<&Ray> = vec![];
    let start_cell = map.get(&start_pos);
    match start_cell.c_type {
        CellType::Empty => next_to_process.push(start_cell.get_potential_ray(dir).unwrap()),
        _ => next_to_process.append(&mut non_empty_cell_next_dirs(start_cell, dir)),
    }

    while let Some(ray) = next_to_process.pop() {
        let mut new_rays = manage_ray_cast(map, ray, &mut solution);
        next_to_process.append(&mut new_rays);
    }

    let mut resulting_marked = Map2D::new(vec![vec![0; map.width()]; map.height()]);
    for to_mark in solution.marks {
        match to_mark {
            ToMark::Horizontal(pos, size) => resulting_marked
                .iter_dir(*pos, Direction::RIGHT, true)
                .take(*size)
                .for_each(|pos_to_mark| *resulting_marked.get_mut(&pos_to_mark) += 1),
            ToMark::Vertical(pos, size) => resulting_marked
                .iter_dir(*pos, Direction::DOWN, true)
                .take(*size)
                .for_each(|pos_to_mark| *resulting_marked.get_mut(&pos_to_mark) += 1),
        }
    }
    if context.is_debug() {
        log!(debug, context, "Result Map\n{}", to_string(&resulting_marked))
    }
    return resulting_marked.iter_all_fast().filter(|pos| *resulting_marked.get(pos) > 0).count();
}

pub fn puzzle(context: &Context, lines: &Vec<String>) {
    let map = parse(lines);
    let result_part1 = solve(&map, Pos { x: 0, y: 0 }, &Direction::RIGHT, context);
    let mut max_result = result_part1;

    for x in 0..map.width() {
        if x != 0 {
            max_result = max_result.max(solve(&map, Pos { x, y: 0 }, &Direction::DOWN, context));
        }
        max_result = max_result.max(solve(&map, Pos { x, y: map.height()-1 }, &Direction::UP, context));
    }

    for y in 1..map.height()-1 {    
        max_result = max_result.max(solve(&map, Pos { x:0, y }, &Direction::RIGHT, context));
        max_result = max_result.max(solve(&map, Pos { x:map.width()-1, y }, &Direction::LEFT, context));
    }


    check_result!(context, [result_part1, max_result], [46, 7951, 51, 8148]);
}
